plenum/server/node.py

```diff
@@ -2411,13 +2422,109 @@ def doStaticValidation(self, request: Request):
 
         self.execute_hook(NodeHooks.POST_STATIC_VALIDATION, request=request)
 
+    def validateTaaAcceptance(self, request: Request):
+        ledger_id = self.ledger_id_for_request(request)
+
+        if not self.ledgerManager.ledger_info(ledger_id).taa_acceptance_required:
+            logger.trace(
+                "{} TAA acceptance passed for request {}: "
+                "not required for ledger id {}"
+                .format(self, request.reqId, ledger_id)
+            )
+            return
+
+        config_req_handler = self.ledger_to_req_handler.get(CONFIG_LEDGER_ID)
+        if not config_req_handler:
+            raise LogicError("Config request handler is missed")
+
+        taa = None
+        taa_data = config_req_handler.get_taa_data()
+        if taa_data is not None:
+            taa, taa_seq_no, taa_txn_time = taa_data
+
+        if not taa:
+            logger.trace(
+                "{} TAA acceptance passed for request {}: "
+                "taa is empty or missed"
+                .format(self, request.reqId)
+            )
+            return
+
+        # TODO not (None or empty)  ???
+        taa_digest = config_req_handler.get_taa_digest()  # TODO test that digest match taa
+        if not taa_digest:  # TODO test
+            raise LogicError(
+                "Txn Author Agreement digest is missed in state for taa: version {}, seq_no {}, txn_time {}"
+                .format(taa[TXN_AUTHOR_AGREEMENT_VERSION], taa_seq_no, taa_txn_time)
+            )
+
+        # TODO INDY-2068
+        taa_aml = {}
+        # taa_aml = config_req_handler.get_taa_aml()
+        # if not taa_aml:
+        #    raise TaaAmlNotSetError(
```

Comment:
I think we can raise a LogicError here, since there is a dynamic validation that TAA can be set only after there is non-empty AML present. So, according to this logic, TAA without an AML can not happen.

Reply:
updated